Implement XFS quota for overlay2

amir73il opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**

Everybody wants to set disk quota limits on their containers.
Devicemapper got support for --storage-opt size, then zfs
and btrfs.

This pull request adds support for --storage-opt size to overlay2.

The semantics of the size parameter are slightly different then those of devmapper/zfs/btrfs.
For the latter, it means total size limit of composite base image + diff.
For overlay, it means size limit for the diff directory.
I chose to use the same parameter to simplify things a bit.

**\- How I did it**

The overlay2 driver (just as well as overlay and aufs) uses directories
to store the file differences from base image.
XFS supports the project quotas feature, which allows setting a disk quota
for a sub directory and its descendants when the directory is created.
The --storage-opt size option to overlay2 utilizes XFS project quotas to
set a hard limit on the disk usage of a container directory.
The option is therefore applicable only when backing fs is xfs and
only when it was mounted with the 'pquota' mount option to enable
project quota support. 

**Why only overlay2?**
Same support could be added to aufs, but it seems to belong to the past,
so I left it there.
overlay (1) cannot use project quotas, because hardlinks are not allowed across
directories that are assigned to different project quotas.

**Why only xfs?**
XFS has had project quotas for a long time. Ext4 is gained project quota support
in kernel v4.5 and the quota control API has been generalized to fit
both XFS and Ext4.
When these kernels get to distributions, the specific XFS project quota code can
be amended to serve overlay2 over both xfs and ext4 backing fs.

**Known issues**
- Driver ignores the /etc/projid configuration file, so it does not coordinate container project id's with the xfs_quota tool.
- Project quota entries are not cleared when container is removed.

**\- How to verify it**

<pre>
# grep xfs /etc/fstab
/dev/mapper/test-xfs /var/lib/docker xfs pquota 0 0

# mount | grep xfs
/dev/mapper/test-xfs on /var/lib/docker type xfs (rw,relatime,attr2,inode64,prjquota)

# dockerd --storage-driver=overlay2 -D&
# docker run --storage-opt size=5g -t -i ubuntu /bin/bash
...
root@3812172791eb:/# df -h .
Filesystem      Size  Used Avail Use% Mounted on
overlay         5.0G   12K  5.0G   1% /
</pre>
 

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Adds support for --storage-opt size to overlay2 over xfs.

**\- A picture of a cute animal (not mandatory but encouraged)**
![cutebunny](https://cloud.githubusercontent.com/assets/453028/16928038/371a360a-4d39-11e6-97ea-dcd3ebbd2361.jpg)

cpuguy83 commented:
Wow pretty cool, thanks for working on this!

crosbymichael commented:
Moving to code review, sorry for the delay.

@amir73il can you rebase this on master?

dmcgowan commented:
Excellent work, this is really cool to see.

From a code layout perspective, I would like to see the xfs specific logic go into a separate file. Ideally we could define an interface for quota management which could apply the quotas on create and be initialized by the storage options. The storage option parsing should be moved to one location, the `xfs.` prefixed options could be collected together and used to initialize the xfs quota manager.

With the possibility of supported ext4 in the near future, how similar to do you think the interface to apply the quotas will be? Having less xfs-specific logic in the main overlay file should help make adding support for other backing fs in the future cleaner if the usage is going to differ significantly.

amir73il commented:
@dmcgowan I fully agree with the code layout comments. Just wanted to put
the code out there before sorting everything out.
Ext4 project quota feature is going to be the same as xfs project quota.
The kernel ABI will be new and common to both ext4 and xfs, but xfs
specific  ioctls will still be needed to support older kernels.
@albamc, can you take the task of moving xfs code to common files and
integrate it with your overlay patch?
I do suggest that you use your inode number approach and drop my projid
assignment code for start.

On Aug 12, 2016 10:02 PM, "Derek McGowan" <notifications@github.com> wrote:

Excellent work, this is really cool to see.

From a code layout perspective, I would like to see the xfs specific logic
go into a separate file. Ideally we could define an interface for quota
management which could apply the quotas on create and be initialized by the
storage options. The storage option parsing should be moved to one
location, the xfs. prefixed options could be collected together and used to
initialize the xfs quota manager.

With the possibility of supported ext4 in the near future, how similar to
do you think the interface to apply the quotas will be? Having less
xfs-specific logic in the main overlay file should help make adding support
for other backing fs in the future cleaner if the usage is going to differ
significantly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/pull/24771#issuecomment-239545786>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAbppNR0-Ta3HGn0Z_x0hOTkIIkTo74zks5qfNE8gaJpZM4JPFUv>
.

Commit:
projectquota: utility class for project quota controls

This class implements XFS project quota controls
for setting quota limits on a newly created directory.
It currently supports the legacy XFS specific ioctls.

Using this class, quota limits per container can be set
by directory based storage drivers (e.g. overlay), when
backing storage is XFS mounted with 'pquota' mount option.

TODO: use generic quota control ioctl FS_IOC_FS{GET,SET}XATTR
      for both xfs/ext4 for kernel version >= v4.5

Signed-off-by: Amir Goldstein <amir73il@aquasec.com>
Signed-off-by: albam.c <albam.c@navercorp.com>

amir73il commented:
@crosbymichael rebased to master. Sorry for the delay.
@dmcgowan I addressed your code review comments about code layout.
Left a TODO comment for adding ext4/xfs project quota for kernels >= v4.5.
Many thanks to @albamc for factoring out the project quota bits into a separate class
and for making the project id allocator persistent.

Commit:
overlay2: add support for --storage-opt size

Allow passing --storage-opt size=X to docker create/run commands
for the `overlay2` graphriver.

The size option is only available if the backing fs is xfs that is
mounted with the `pquota` mount option.
The user can pass any size less then the backing fs size.

Signed-off-by: Amir Goldstein <amir73il@aquasec.com>

dmcgowan commented on daemon/graphdriver/projectquota.go:
```diff
@@ -0,0 +1,292 @@
+		return "", err
+	}
+
+	backingFsDev := path.Join(home, "backingFsDev")
```
can we make this name something more meaningful

amir73il replied:
How about backingFsBlockDev? that's all it is, an easy handle to perform the set/get quota limit ioctls, which are performed on global fs level, not on directory level.

dmcgowan commented on daemon/graphdriver/projectquota.go:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	backingFsDev := path.Join(home, "backingFsDev")
+	syscall.Unlink(backingFsDev)
```
Does this need to be removed every time or is this just to guarantee the device node is correct in case the home directory device changes?

amir73il replied:
It is just a guarantee. The way xfs_quota tool does it to get the xfs mount point as argument and deduce the the block device from /proc/mounts. So having the backingFsDev actual node (as a duplicate of the original node (e.g. /dev/dm-1) is just a matter of convenience and less complex code.

dmcgowan commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -277,6 +294,17 @@ func (d *Driver) Create(id, parent, mountLabel string, storageOpt map[string]str
+			return err
+		}
+		// enforce container disk project quota
+		if err := d.quotaCtl.SetQuota(dir, driver.options.quota); err != nil {
```
Could protect this block in `if driver.options.quota.Size > 0` and enter the outer block `if len(storageOpt) > 0`

amir73il replied:
I guess you are right. I copied the parseStorageOpt() chunk from btrfs.go which does the same, but I can change it if you think it is important.

dmcgowan replied:
I don't think it is important, but if you do push changes again I wouldn't mind seeing it changed :)

dmcgowan commented on daemon/graphdriver/projectquota.go:
```diff
@@ -0,0 +1,292 @@
+
+// NewQuotaCtl - initialize project quota support by finding the first project id
+// to be used for the next container create
+func NewQuotaCtl(basePath string) (*QuotaCtl, error) {
```
At what point does this function determine that project quota is not supported. I was able to get it work with pquota option set, but then when I remounted without it there was no error on startup, only an error on container creation.

amir73il replied:
The first thing that NewQuotaCtl() does is to call getMaxProjectID() and the first thing it does is check the project id on the home directory. This test will silently fail and cause projectQuotaSupported = false if XFS does not support project quotas.
This is not an error, because running overlay over xfs without project quota support is perfectly fine. Using the --size option on such a backfs is the error.
I guess I should add a comment about it.

dmcgowan commented on daemon/graphdriver/projectquota.go:
```diff
@@ -0,0 +1,292 @@
+	//
+	// get max project id for next use
+	//
+	maxProjectID, err := getMaxProjectID(basePath)
```
Ideally we could avoid the directory scanning used by this function when pquota is not supported.

amir73il replied:
We do not get to directory scanning, because the first thing that getMaxProjectID() does is get the project id of the home directory. This test will fail if no pquota support.
If pquota is supported and home directory does not have a project id the check will succeed and return value of 0.
There is also a 'hidden' feature here - project quota can be used to set a quota limit (using xfs_quota tool) on the entire storage driver directory (/var/lib/docker/overlay2), to set it apart from the rest of the system. In that case, the root project id will be used as a "start offset" (e.g. echo docker:999 > /etc/projid) and all containers will be assigned project ids > 1000.
This is a way to prevent xfs_quota management from conflicting with docker.
I did not document this feature because I did not want to complicate. It can be documented later on in advanced documentation.

dmcgowan commented on daemon/graphdriver/projectquota.go:
```diff
@@ -0,0 +1,292 @@
+	}
+	fsx.fsx_projid = C.__u32(projectID)
+	fsx.fsx_xflags |= C.FS_XFLAG_PROJINHERIT
+	_, _, errno = syscall.Syscall(syscall.SYS_IOCTL, getDirFd(dir), C.FS_IOC_FSSETXATTR,
```
Will this show up as an xattr on the directory?

amir73il replied:
No. This is not an extended attribute. It is an xfs internal inode flag (xflags). You can see them as well as the project id with the command:
xfs_io -c stat <path>

dmcgowan commented:
I was able to test this and verify the quotas work as expected. The detection of the feature does not seem to be working. Might be helpful to print the error on quota ctl creation at debug level as well.

amir73il commented:
@dmcgowan thanks for the review. I addressed most of the comments by adding more documentation in the code. Regrading your statement that  "The detection of the feature does not seem to be working", I am not sure what you mean. overlay2 driver automatically detects if pquota is enabled at Init() and prints a debug message with the boolean value of 'projectQuotaSupported'. If projectQuotaSupported is true, then --storage-opt can be used when creating containers. Otherwise (either because not xfs or not pquota), an error is generated on container create:
--storage-opt is supported only for overlay over xfs with 'pquota' mount option

What you described is how it was intended to work. Please let me know if I am missing something.

dmcgowan commented:
@amir73il did another round of testing. The issue I ran into only occurred when I mounted the device with pquota option, then re-mounted without it. When I used a freshly formatted device I did not run into the issue. I thought I tested yesterday I tried with a fresh device, but since my testing today shows it working I don't see it is a problem.

dmcgowan commented:
I did a few more rounds of testing and created a script to help capture the results. Note I am using a loopback device to do the testing and fresh xfs created with default options.

My test script `test_24771.sh` can be found here https://gist.github.com/dmcgowan/a64d67a6ac067f40598a3ce9473c0823

Without pquota mount option (expected projectQuotaSupported=false)
```
# ./test_24771.sh t1
meta-data=/dev/loop0             isize=512    agcount=4, agsize=32768 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=0
data     =                       bsize=4096   blocks=131072, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=855, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
/dev/loop0 on /tmp/tmp.8VQNWZEUOc type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
... removed irrelevant output ...
DEBU[0001] [graphdriver] trying provided driver: overlay2 
DEBU[0001] getMaxProjectID(/tmp/tmp.8VQNWZEUOc/overlay2) = 1 
DEBU[0001] backingFs=xfs,  projectQuotaSupported=true   

$ docker run -it --storage-opt "size=30MB" alpine /bin/sh
... removed irrelevant output ...
docker: Error response from daemon: Failed to set quota limit for projid 1 on /tmp/tmp.idgFjJZAS7/overlay2/backingFsBlockDev: function not implemented.
```

With pquota mount option (works as expected)
```
# ./test_24771.sh t2 -o pquota
meta-data=/dev/loop0             isize=512    agcount=4, agsize=32768 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=0
data     =                       bsize=4096   blocks=131072, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=855, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
/dev/loop0 on /tmp/tmp.Ovy1vkiAi6 type xfs (rw,relatime,seclabel,attr2,inode64,prjquota)
... removed irrelevant output ...   
DEBU[0001] [graphdriver] trying provided driver: overlay2 
DEBU[0001] getMaxProjectID(/tmp/tmp.Ovy1vkiAi6/overlay2) = 1 
DEBU[0001] backingFs=xfs,  projectQuotaSupported=true

$ docker run -it --storage-opt "size=30MB" alpine /bin/sh
... removed irrelevant output ...
/ # dd if=/dev/zero of=testfile bs=1M count=64
dd: writing 'testfile': No space left on device
31+0 records in
29+1 records out
```

amir73il commented:
@dmcgowan I see, so I will need to try to set a quota on a temp dir for a correct test of pquota support.
I think I also ran my test after remount and not after a fresh mkfs it did passed, strange...
Thanks for the test script. I will look into it

dmcgowan commented:
@amir73il it was happening on non-temp directories prior to making the script, didn't know if using a loopback had an effect. I would also be ok with an `overlay2.allowquota` options or something like that to explicitly turn it on through the daemon storage options if there is no fast and simple way to detect the flag on Init.

albamc commented on daemon/graphdriver/projectquota.go:
```diff
@@ -0,0 +1,310 @@
+	//
+	var d C.fs_disk_quota_t
+	d.d_version = C.FS_DQUOT_VERSION
+	d.d_id = fsx.fsx_projid
```
I think fsx.fsx_projid is can be 0 because it wasn't initialized.
Am I wrong ?

amir73il commented:
Oops! Good catch :/

amir73il commented:
@dmcgowan I rather not have options that the user doesn't need to know about (i.e. overlay2.allowquota) 
I pushed a fix with proper detection by setting quota on a test dir. Also addressed your other review comments and tested with your test script. Thanks!
I also ran some tests trying to mount -o remount,pquota and it seems to be impossible to effectively turn pquota on while xfs is mounted. Even when mount options seem to have changed on remount on my host, the mount options inside my docker development container are still 'noquota'.
The only way I could get quota support is by mounting xfs with pquota to begin with and then running docker daemon (like your test script does).

dmcgowan commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -150,11 +161,33 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
+		// Try to enable project quota support over xfs.
+		if d.quotaCtl, err = graphdriver.NewQuotaCtl(home); err == nil {
+			//
+			// Test if file system supports project quotas by trying to set a quota
```
Ideally this check would be part of `NewQuotaCtl` and done before `getMinProjectID` is called. My worry is that for xfs user who aren't using this feature, they will see a slower start-up linear to the number of layers they have. Do you have any numbers of how long the scan takes?

amir73il replied:
I understand your concern. scan should be just as fast as ls -l /var/lib/docker/overlay2 > /dev/null.
but anyway, I pushed a change as you requested.

dmcgowan commented:
Latest change looks good and working now as I would expect. Before approving 2 more questions...
- Is it possible to have a reserved project id (maybe 1) for testing whether pquota is supported? The benefit would be saving the need figure out the next project id and could just assume that 1 is always available for testing.
- Can the project id only be set on a directory or could it also be set on `backingFsBlockDev`? This only real benefit of this would be so a this test directory which no other use doesn't show up in the logs.

@crosbymichael any other changes you would like to see?

amir73il commented:
@dmcgowan I did "reserve" a special project id (1) for the test. But it is only reserved in the docker project id allocation range. In the example I wrote in the comment above NewQuotaCtl, /var/lib/docker/overlay2 is assigned the id 999. Then on every daemon start it will create quotatest dir and assign it to projid 1000 and set quota for projid 1000. All container will use projid > 1000.
This can be documented in the system's /etc/projid file, e.g.:
  docker:999
  dockertest:1000
  # >1000 reserved for docker containers

There is no real way to actually reserve a projid so others won't use it,
but since the most common way to configure project quotas is by xfs_quota tool,
"reserving" in its /etc/projid configuration file seems good enough.

Furthermore, it is possible to set a quota for projid X regardless if there are any directory
assigned to this id at all and directories can be assigned to project id X later on.

amir73il commented:
@dmcgowan to answer your second question, no there is no need to create the testdir. I can test set the quota on minProjectID+1 on the blockdev without creating testdir. there is really no need for it.

dmcgowan commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -76,15 +77,25 @@ const (
-var backingFs = "<unknown>"
+var (
+	backingFs             = "<unknown>"
+	projectQuotaSupported = false
```
If I stare at the code long enough I suppose I am bound to find more nits, this variable is not necessary since quota support can be tested with `d.quotaCtl != nil`

amir73il replied:
True, but IMO it doesn't hurt to have a human readable boolean for that.

dmcgowan reviewed (approved):
Tested once again, everything is working!

LGTM

domdom82 commented:
Awesome! Any chance we get the ext4 version of this in 1.13, too?

amir73il commented:
@domdom82 what's the rush to get the ext4 version in 1.13? ext4 project quota support has only been merged since kernel version 4.5 and has had some stability fixes in following releases. The newest LTS kernel version used by distros in 4.4. So who is to gain from this?

domdom82 commented:
@amir73il yes, I believe I saw the most recent fixes on ext4 prj quota in 4.8 here https://github.com/domdom82/linux/commit/079788d01e7ba9d7366d7bd2a0db9cab5944e85b
However given that the logic in Docker should be fairly similar for both fs and the widespread use of ext4 and the growing demand for that feature I figured it'd be great to get it early on. But I agree we should not rush it and sacrifice stability so 1.13 may be a bit too early indeed.

thaJeztah commented:
ping @crosbymichael PTAL!

domdom82 commented:
anything still blocking this?

crosbymichael commented:
LGTM

runcom commented:
This broke compilation on Fedora (24 for me)

runcom commented:
```
# github.com/docker/docker/daemon/graphdriver
.gopath/src/github.com/docker/docker/daemon/graphdriver/projectquota.go:20:8: error: redefinition of 'struct fsxattr'
 struct fsxattr {
        ^~~~~~~
In file included from .gopath/src/github.com/docker/docker/daemon/graphdriver/projectquota.go:17:0:
/usr/include/linux/fs.h:155:8: note: originally defined here
 struct fsxattr {
        ^~~~~~~
cp: cannot stat 'bundles/latest/dynbinary-client/docker': No such file or directory
cp: cannot stat 'bundles/latest/dynbinary-daemon/dockerd': No such file or directory
now call sudo systemctl restart docker
```
@amir73il PTAL

amir73il commented:
@runcom sorry for that breakage. Please try this patch and let me know if it fixes your problem.
https://github.com/aquasecurity/docker/commit/4de24fedea8f99230789d56a9c8c7b0c278a039a

dmcgowan commented:
Fixes build for me (Fedora 23 4.7.6)

amir73il commented:
Sent PR #27459

dnephin commented:
This broke manpages generation.  The `daemon/graphdriver` package is (indirectly) imported from `cli/command`. Previously the client package did not require `linux/fs.h` to compile.

I think we should add the man page generation as a build to CI to catch these problems. That will be much easier with something like #27359.

For now, can we move `daemon/graphdriver/projectquota.go` to either `daemon/graphdriver/overlay2/` or `daemon/graphdriver/quota` ?

ghostplant commented on docs/reference/commandline/run.md:
```diff
@@ -198,8 +198,13 @@ The `-w` lets the command being executed inside directory given, here
     $ docker run -it --storage-opt size=120G fedora /bin/bash
 
 This (size) will allow to set the container rootfs size to 120G at creation time.
-User cannot pass a size less than the Default BaseFS Size. This option is only
```
According to the Doc, I am a little confused.
For xfs + overlay2 quota control, whether container rootfs size is limited to 120G, or the top layer is limited to 120G?

amir73il replied:
Only the top layer.
So only the size of changes from root is limited, including the entire size of all copies up files.

ghostplant replied:
How about zfs + overlay2?

amir73il replied:
See below: For the `overlay2` storage driver, the size option is only available if the backing fs is `xfs`. You cannot use the size option for xfs + overlay2

amir73il replied:
I meant you cannot use size option for xfs + overlay2 of course.

amir73il replied:
Oef. Autocorrect. zfs + overlay2 size option is not available

ghostplant replied:
@amir73il Is it needed to consider removing quota settings when the layer is removed?

ghostplant commented:
Hi, have the known issues solved?

amir73il commented:
Both issues are not relevant for systems that do not use xfs project quotas
management.

Both issues can be addressed using the following instructions
(from the comment in projectquota.go):

xfs_quota tool can be used to assign a project id to the driver home
directory, e.g.:
echo 999:/var/lib/docker/overlay2 >> /etc/projects
echo docker:999 >> /etc/projid
xfs_quota -x -c 'project -s docker' /<xfs mount point>
In that case, the home directory project id will be used as a "start
offset" and all containers will be assigned larger project ids (e.g. >=
1000). This is a way to prevent xfs_quota management from conflicting with
docker.

ghostplant commented:
Got it. Thanks!

rhvgoyal commented:
This sounds very useful. 

I guess something like this might be useful for limiting volume size also.

cmurf commented:
Would Docker be able to use storaged for configuring quotas, if it were available? All the file systems have different quota features and interfaces, it'd be nice to see their overlapping quota support abstracted so Docker can use (project/directory/subvolume) quotas regardless of filesystem.

amir73il commented:
@amir73il <https://github.com/amir73il> Is it needed to consider removing
quota settings when the layer is removed?

Eventually that should be considered so project IDs could be reused
efficiently. There are only 32bit of projid by default in xfs.
However, current implementation doesn't have a smart proij allocator. It
always uses max projid of existing layers +1 for new layer.
Note that if all layers are removed (or all new layers since projid x) then
on next daemon start those removed layers projids will be reused.

Levovar commented:
I would like to have a question regarding the content of this request: am I right that the implemented change only adds overlay2fs size quota capability to the docker create and docker run interfaces?
Or the same option is available even as a dockerd storage-driver option?

I guess not, because the dockerd documentation page still only mentions "override_kernel_check" as the only overlayfs2 specific parameter.
Is it considered to make this option generic, so cluster administrators do not need to separately specify the quota parameter for each and every container creation command? If we consider that the direct docker create/run interfaces are sometimes "hidden" in an orchestrated cluster -e.g. in a K8 cluster-, the only way to impose generic, cluster-wide quotas would be through the configuration of dockerd itself (like with devicemapper).

amir73il commented:
@Levovar you are right that this PR only adds storage-opt size to create/run commands.

IMO, it should not be a problem to add a dockerd generic storage-opt default_size so any create/run command that don't specify storage-opt size should use this default size (but I am not volunteering to implement it).

However, this requirement is not specific to overlay2 storage driver.
There is no such feature for btrfs (btrfs.min_space is quite different).
You may think that dm.basesize option is the same as default_size, but it is not.
dm.basesize is more like max_size, because you cannot create any container with
storage size larger then dm.basesize.
So default_size option may be useful for devicemapper as well, for example,
when you want to set the maximum container size to 50G, but set a default quota
for containers that are created without specifying storage size.

imkin commented:
@amir73il I had asked for something similar here https://github.com/docker/docker/issues/30491

imkin commented:
@amir73il I was planning to have a PR and would love to have your eyes on the changes. PR would be to fix https://github.com/docker/docker/issues/30491 and which fits in line with your comment above.
